File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i;
    int find(int x,vector<int>in){
        for(int i=0;i<in.size();i++){
            if(in[i]==x) return i;
        }
        return -1;
    }
    TreeNode* fun(vector<int>&pre,vector<int>&in,int start,int end){
        if(start>end) return NULL;
        TreeNode* root=new TreeNode(pre[i]);
        
        int idx=find(pre[i],in);
        i++;
        if(idx==-1) return NULL;
        root->left=fun(pre,in,start,idx-1);
        root->right=fun(pre,in,idx+1,end);
        return root;

       
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        i=0;
        int n=preorder.size();
        return fun(preorder,inorder,0,n-1);
        
        
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.i = 0

    def fun(self, start, end, pre, ino):
        if start > end:
            return None

        node = TreeNode(pre[self.i])
        root = pre[self.i]
        self.i += 1

        pos = -1
        for j in range(start, end + 1):
            if ino[j] == root:
                pos = j
                break

        node.left = self.fun(start, pos - 1, pre, ino)
        node.right = self.fun(pos + 1, end, pre, ino)

        return node

    def buildTree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        self.i = 0
        n = len(preorder)

        return self.fun(0, n - 1, preorder, inorder)
        